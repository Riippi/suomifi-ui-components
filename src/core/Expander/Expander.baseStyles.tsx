import { css } from 'styled-components';
import { withSuomifiTheme, TokensAndTheme } from '../theme';
import { button, element, font } from '../theme/reset';
import { absolute } from '../../utils/css';
import { padding, absoluteFocus } from '../theme/utils';

import { ExpanderProps } from './Expander';

export const baseStyles = withSuomifiTheme(
  ({ theme }: TokensAndTheme & Partial<ExpanderProps>) => {
    return css`
  ${element({ theme })}
  ${font({ theme })('bodyText')}
  background-color: ${theme.colors.whiteBase};
  ${absolute('before')}
  position: relative;
  padding: 0;
  border-radius: ${theme.radius.basic};
  box-shadow: ${theme.shadows.panelShadow};

  &:before {
    background-color: ${theme.colors.highlightLight4};
    opacity: 0;
  }

  &.fi-expander--open:before {
    opacity: 1;
    transition: opacity ${`${theme.transitions.basicTime}
      ${theme.transitions.basicTimingFunction}`};
  }

  & .fi-expander_title {
    ${button({ theme })}
    position: relative;
    display: block;
    width: 100%;
    font-size: ${theme.typography.bodySemiBold};
    min-height: 60px;

    &:focus {
      outline: 0;
      &:after {
        ${absoluteFocus}
      }
    }

    &--no-tag {
      padding: 17px ${theme.spacing.xxxl} 16px ${theme.spacing.m};
      color: ${theme.colors.highlightBase};
    }
  }

  & .fi-expander_title-icon {
    position: absolute;
    height: 20px;
    width: 20px;
    top: 0;
    right: 0;
    margin: ${theme.spacing.m};
  }
  & .fi-expander_title--open .fi-expander_title-icon,
  & .fi-expander_title-icon--open {
    transform: rotate(-180deg);
  }

  & > .fi-expander_content {
    position: relative;
    display: block;
    height: 0;
    overflow: hidden;
    word-break: break-word;
    transform: scaleY(0);
    transform-origin: top;
    transition: all ${`${theme.transitions.basicTime}
      ${theme.transitions.basicTimingFunction}`};
    will-change: transition, height;
    &:not(.fi-expander_content--no-padding) {
      padding: 0 ${theme.spacing.insetXl};
    }
    &.fi-expander_content--open {
      height: 10%;
      overflow: visible;
      /* This is very robust - cannot animate dynamic height with height-definition */
      animation: fi-expander_content-anim ${theme.transitions.basicTime}
        ${theme.transitions.basicTimingFunction} 1 forwards;
      &:not(.fi-expander_content--no-padding) {
        ${padding({ theme })('0', 'm', 'm', 'm')}
      }
    }
    @keyframes fi-expander_content-anim {
      0% {
        height: auto;
        transform: scaleY(0);
      }
      100% {
        transform: scaleY(1);
      }
    }
  }
`;
  },
);
