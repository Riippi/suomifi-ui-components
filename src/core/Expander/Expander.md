```jsx
import { Expander } from 'suomifi-ui-components';

<Expander title="Test expander" className="expander-test">
  Test expander
</Expander>;
```

```jsx
import { Expander } from 'suomifi-ui-components';

<Expander.group OpenAll="Open all" CloseAll="Close all">
  <Expander title="Test expander 1">Test expander content 1</Expander>
  <Expander title="Test expander 2">Test expander content 2</Expander>
  <Expander title="Test expander 3">Test expander content 3</Expander>
</Expander.group>;
```

## Controlled

- State for the individual Expanders are stored outside of the component and user has full control.
- Therefore when clicking the individual Expander they are not opened by default, user have to give the logic to change it.
- It's user's responsibility to keep the state stored outside to be updated as Open/Close All is used.
- `defaultOpen` prop will not work when Expander is in controlled state == `open` prop is given.

```jsx
import { Expander } from 'suomifi-ui-components';

const [expanderThreeOpen, setExpanderThreeOpen] = React.useState(
  false
);

<>
  <Expander.group OpenAll="Open all" CloseAll="Close all">
    <Expander title="Test expander 1">
      Test expander content 1
    </Expander>
    <Expander title="Test expander 2">
      Test expander content 2
    </Expander>
    <Expander
      title="Test expander 3"
      open={expanderThreeOpen}
      onClick={({ openState }) => {
        if (window.confirm('Toggle Expander 3')) {
          setExpanderThreeOpen(!openState);
        }
      }}
    >
      Test expander content 3
    </Expander>
  </Expander.group>
</>;
```
