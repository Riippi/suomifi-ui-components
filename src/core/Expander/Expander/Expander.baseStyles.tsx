import { css } from 'styled-components';
import { defaultThemeTokens as theme } from '../../theme';
import { element, font } from '../../theme/reset';

export const baseStyles = css`
  ${element({ theme })}
  ${font({ theme })('bodyText')}
    background-color: ${theme.colors.whiteBase};
  position: relative;
  padding: 0;
  border-radius: ${theme.radius.basic};
  box-shadow: ${theme.shadows.panelShadow};
  display: block;
  width: 100%;
  max-width: 100%;

  &:before {
    background-color: ${theme.colors.highlightLight4};
    opacity: 0;
  }
`;
