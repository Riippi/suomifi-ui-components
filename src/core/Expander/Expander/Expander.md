```jsx
import {
  Expander,
  ExpanderTitleButton,
  ExpanderContent
} from 'suomifi-ui-components';

<Expander className="expander-test" defaultOpen={true}>
  <ExpanderTitleButton>Test expander</ExpanderTitleButton>
  <ExpanderContent>Test expander</ExpanderContent>
</Expander>;
```

```jsx
import {
  Expander,
  ExpanderTitle,
  ExpanderContent,
  Checkbox
} from 'suomifi-ui-components';

<Expander className="expander-test" defaultOpen={true}>
  <ExpanderTitle>
    <Checkbox hintText="Checkbox hint text">Checkbox label</Checkbox>
  </ExpanderTitle>
  <ExpanderContent>Test expander</ExpanderContent>
</Expander>;
```

```jsx
import {
  Expander,
  ExpanderGroup,
  ExpanderTitleButton,
  ExpanderContent
} from 'suomifi-ui-components';

<ExpanderGroup
  OpenAllText="Open all"
  AriaOpenAllText="Open all expanders"
  CloseAllText="Close all"
  AriaCloseAllText="Close all expanders"
>
  <Expander>
    <ExpanderTitleButton>Test expander 1</ExpanderTitleButton>
    <ExpanderContent>Test expander content 1</ExpanderContent>
  </Expander>
  <Expander>
    <ExpanderTitleButton>Test expander 2</ExpanderTitleButton>
    <ExpanderContent>Test expander content 2</ExpanderContent>
  </Expander>
  <Expander>
    <ExpanderTitleButton>Test expander 3</ExpanderTitleButton>
    <ExpanderContent>Test expander content 3</ExpanderContent>
  </Expander>
</ExpanderGroup>;
```

## Controlled

- State for the individual Expanders are stored outside of the component and user has full control.
- Therefore when clicking the individual Expander they are not opened by default, user have to give the logic to change it.
- It's user's responsibility to keep the state stored outside to be updated as Open/Close All is used.
- `defaultOpen` prop will not work when Expander is in controlled state == `open` prop is given.

```jsx
import {
  Expander,
  ExpanderGroup,
  ExpanderTitleButton,
  ExpanderContent
} from 'suomifi-ui-components';

const [expanderThreeOpen, setExpanderThreeOpen] = React.useState(
  false
);

<>
  <ExpanderGroup
    OpenAllText="Open all"
    AriaOpenAllText="Open all expanders"
    CloseAllText="Close all"
    AriaCloseAllText="Close all expanders"
  >
    <Expander>
      <ExpanderTitleButton>Test expander 1</ExpanderTitleButton>
      <ExpanderContent>Test expander content 1</ExpanderContent>
    </Expander>
    <Expander>
      <ExpanderTitleButton>Test expander 2</ExpanderTitleButton>
      <ExpanderContent>Test expander content 2</ExpanderContent>
    </Expander>
    <Expander
      open={expanderThreeOpen}
      onOpenChange={(open) => {
        if (window.confirm('Toggle Expander 3')) {
          setExpanderThreeOpen(!open);
        }
      }}
    >
      <ExpanderTitleButton>Test expander 3</ExpanderTitleButton>
      <ExpanderContent>Test expander content 3</ExpanderContent>
    </Expander>
  </ExpanderGroup>
</>;
```
