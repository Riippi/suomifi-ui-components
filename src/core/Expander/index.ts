export { Expander, ExpanderProps } from './Expander/Expander';
export {
  ExpanderGroup,
  ExpanderGroupProps,
} from './ExpanderGroup/ExpanderGroup';
export {
  ExpanderContent,
  ExpanderContentProps,
} from './ExpanderContent/ExpanderContent';
export {
  ExpanderTitleButton,
  ExpanderTitleButtonProps,
} from './ExpanderTitle/ExpanderTitleButton';
