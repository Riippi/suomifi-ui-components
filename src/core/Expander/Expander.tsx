import React, { Component, Fragment } from 'react';
import { default as styled } from 'styled-components';
import { withSuomifiDefaultProps } from '../theme/utils';
import { TokensProp, InternalTokensProp } from '../theme';
import { baseStyles } from './Expander.baseStyles';
import {
  Expander as CompExpander,
  ExpanderProps as CompExpanderProps,
} from '../../components/Expander/Expander';
import { Icon } from '../Icon/Icon';
import classnames from 'classnames';
import {
  ExpanderGroupProps,
  ExpanderGroup,
  ExpanderGroupConsumer,
} from './ExpanderGroup';

const iconClassName = 'fi-expander_title-icon';
const iconOpenClassName = `${iconClassName}--open`;
const noPaddingClassName = `fi-expander_content--no-padding`;

type ExpanderVariant = 'expander' | 'expanderGroup';

export interface ExpanderProps extends CompExpanderProps, TokensProp {
  /** Remove padding from expandable content area (for background usage with padding in given container etc.) */
  noPadding?: boolean;
  /**
   * 'expander' | 'expanderGroup'
   * @default expander
   */
  variant?: ExpanderVariant;
}

const StyledExpander = styled(
  ({ tokens, noPadding, ...passProps }: ExpanderProps & InternalTokensProp) => {
    return (
      <CompExpander
        {...passProps}
        contentProps={{
          className: classnames({ [noPaddingClassName]: noPadding }),
        }}
      />
    );
  },
)`
  ${(props) => baseStyles(props)};
`;

interface ExpanderState {
  openState: boolean;
}

/**
 * <i class="semantics" />
 * Used for openable expander
 */
class ExpanderItem extends Component<ExpanderProps> {
  static group = (props: ExpanderGroupProps) => {
    return <ExpanderGroup {...withSuomifiDefaultProps(props)} />;
  };

  /** State is only used to update the caret-icon */
  state: ExpanderState = {
    openState:
      this.props.defaultOpen !== undefined ? this.props.defaultOpen : false,
  };

  componentDidUpdate(prevProps: ExpanderProps, prevState: ExpanderState) {
    if (
      !!this.props.consumer &&
      !!prevProps.consumer &&
      this.props.consumer.toggleAllExpanderState !==
        prevProps.consumer.toggleAllExpanderState
    ) {
      const { openState } = this.state;
      const {
        expanderGroup,
        index,
        consumer: { toggleAllExpanderState },
        open,
      } = this.props;

      if (
        !!expanderGroup &&
        index !== undefined &&
        ((open === undefined &&
          !!openState !== toggleAllExpanderState.toState) ||
          (open !== undefined && !!open !== toggleAllExpanderState.toState))
      ) {
        this.handleClick();
      }
    }

    if (
      prevState.openState !== this.state.openState ||
      prevProps.open !== this.props.open
    ) {
      const {
        expanderGroup,
        index,
        consumer: { onClick: consumerOnClick } = { onClick: undefined },
      } = this.props;
      if (!!expanderGroup && !!consumerOnClick && index !== undefined) {
        consumerOnClick(index);
      }
    }
  }

  handleClick = () => {
    const { open, onClick } = this.props;
    const { openState } = this.state;
    const controlled = open !== undefined;
    const newOpenState = controlled ? !!open : !openState;

    this.setState({ openState: newOpenState });
    if (!!onClick) {
      onClick({ openState: newOpenState });
    }
  };

  render() {
    const {
      variant = 'expander',
      open,
      title,
      titleTag,
      index,
      consumer,
      ...passProps
    } = withSuomifiDefaultProps(this.props);
    const openState = open !== undefined ? !!open : this.state.openState;

    if (variant === 'expanderGroup' && 'openAll' in passProps) {
      return <ExpanderGroup {...(passProps as ExpanderGroupProps)} />;
    }

    return (
      <StyledExpander
        {...passProps}
        index={index}
        onClick={this.handleClick}
        open={open}
        titleTag={titleTag}
        title={
          <Fragment>
            {title}
            {!titleTag && (
              <Icon
                icon="chevronDown"
                className={classnames(iconClassName, {
                  [iconOpenClassName]: openState,
                })}
                color={passProps.tokens.colors.highlightBase}
              />
            )}
          </Fragment>
        }
      />
    );
  }
}

export class Expander extends Component<ExpanderProps> {
  static group = (props: ExpanderGroupProps) => {
    return <ExpanderGroup {...withSuomifiDefaultProps(props)} />;
  };

  render() {
    return !!this.props.expanderGroup ? (
      <ExpanderGroupConsumer>
        {(consumer) => <ExpanderItem {...this.props} consumer={consumer} />}
      </ExpanderGroupConsumer>
    ) : (
      <ExpanderItem {...this.props} />
    );
  }
}
