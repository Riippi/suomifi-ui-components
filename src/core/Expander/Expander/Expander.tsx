import React, { Component, ReactNode } from 'react';
import { default as styled } from 'styled-components';
import classnames from 'classnames';
import { withSuomifiDefaultProps } from '../../theme/utils';
import { AutoId } from '../../../utils/AutoId';
import { TokensProp, InternalTokensProp } from '../../theme';
import { HtmlDiv } from '../../../reset';
import { baseStyles } from './Expander.baseStyles';
import {
  ExpanderGroupConsumer,
  ExpanderGroupProviderState,
} from '../ExpanderGroup/ExpanderGroup';

const baseClassName = 'fi-expander';
const openClassName = `${baseClassName}--open`;

export interface ExpanderProviderState {
  /** Callback for communicating ExpanderTitle button event to Expander  */
  onToggleExpander: () => void;
  /** Open state for expander */
  open: boolean;
  /** Id for expander button */
  titleId: string | undefined;
  /** Id for expander content */
  contentId: string | undefined;
}

const defaultProviderValue: ExpanderProviderState = {
  onToggleExpander: () => null,
  open: false,
  titleId: undefined,
  contentId: undefined,
};

const {
  Provider: ExpanderProvider,
  Consumer: ExpanderConsumer,
} = React.createContext(defaultProviderValue);

interface InternalExpanderProps {
  /**
   * Children, extend type ExpanderTitleBaseProps or ExpanderContentBaseProps
   * ExpanderProviderState context is used to communicate between title, content and expander
   */
  children: ReactNode;
  /** Custom classname to extend or customize */
  className?: string;
  /**
   * Id for expander, must be unique. Duplicate id's break ExpanderGroup functionality.
   * Autogenerated if not provided
   */
  id?: string;
  /** Default status of expander open
   * @default false
   */
  defaultOpen?: boolean;
  /** Controlled open property */
  open?: boolean;
  /** Event handler to execute when clicked */
  onOpenChange?: (open: boolean) => void;
  consumer?: ExpanderGroupProviderState;
}

export interface ExpanderProps extends InternalExpanderProps, TokensProp {}

export interface ExpanderTitleBaseProps {
  /** Custom classname to extend or customize */
  className?: string;
  /** Expander consumer for open state and toggle open callback */
  consumer: ExpanderProviderState;
}

export interface ExpanderContentBaseProps {
  /** Custom classname to extend or customize */
  className?: string;
  /** Expander consumer for open state */
  consumer: ExpanderProviderState;
}

interface BaseExpanderProps extends InternalExpanderProps {
  id: string;
}

class BaseExpander extends Component<BaseExpanderProps> {
  state: ExpanderState = {
    openState: this.props.defaultOpen || false,
  };

  constructor(props: BaseExpanderProps) {
    super(props);
    if (!!props.consumer) {
      const defaultOpen =
        props.open !== undefined ? props.open : props.defaultOpen || false;
      props.consumer.onExpanderOpenChange(props.id, defaultOpen);
    }
  }

  componentDidUpdate(prevProps: ExpanderProps, prevState: ExpanderState) {
    const { consumer, open } = this.props;
    const controlled = open !== undefined;
    if (
      !!consumer &&
      prevProps.id !== undefined &&
      prevProps.id !== this.props.id
    ) {
      consumer.onExpanderOpenChange(prevProps.id, undefined);
    }
    if (
      !!consumer &&
      consumer.toggleAllExpanderState !==
        prevProps.consumer?.toggleAllExpanderState
    ) {
      if (
        (!controlled &&
          !!this.state.openState !== consumer.toggleAllExpanderState.toState) ||
        (controlled && open !== consumer.toggleAllExpanderState.toState)
      ) {
        this.handleOpenChange();
      }
    }
    if (
      (!controlled && prevState.openState !== this.state.openState) ||
      (controlled && prevProps.open !== open)
    ) {
      if (!!consumer && this.props.id !== undefined) {
        const currentState = controlled ? !!open : this.state.openState;
        consumer.onExpanderOpenChange(this.props.id, currentState);
      }
    }
  }

  componentWillUnmount() {
    if (!!this.props.consumer && !!this.props.consumer.onExpanderOpenChange) {
      this.props.consumer.onExpanderOpenChange(this.props.id, undefined);
    }
  }

  handleOpenChange = () => {
    const { open, onOpenChange } = this.props;
    const { openState } = this.state;
    const controlled = open !== undefined;
    const newOpenState = controlled ? !!open : !openState;
    if (!controlled) {
      this.setState({ openState: newOpenState });
    }
    if (!!onOpenChange) {
      onOpenChange(newOpenState);
    }
  };

  render() {
    const {
      id,
      open,
      defaultOpen,
      onOpenChange,
      className,
      children,
      consumer,
      ...passProps
    } = this.props;
    const openState = open !== undefined ? !!open : this.state.openState;
    const titleId = `${id}_title`;
    const contentId = `${id}_content`;

    return (
      <HtmlDiv
        id={id}
        {...passProps}
        className={classnames(className, baseClassName, {
          [openClassName]: !!openState,
        })}
      >
        <ExpanderProvider
          value={{
            open: openState,
            contentId,
            titleId,
            onToggleExpander: this.handleOpenChange,
          }}
        >
          {children}
        </ExpanderProvider>
      </HtmlDiv>
    );
  }
}

const StyledExpander = styled(
  ({
    tokens,
    id: propId,
    ...passProps
  }: ExpanderProps & InternalTokensProp) => (
    <AutoId id={propId}>
      {(id) => <BaseExpander id={id} {...passProps} />}
    </AutoId>
  ),
)`
  ${(props) => baseStyles(props)};
`;

interface ExpanderState {
  openState: boolean;
}

/**
 * <i class="semantics" />
 * Hide or show content with always visible title
 */
export class Expander extends Component<ExpanderProps> {
  render() {
    return (
      <ExpanderGroupConsumer>
        {(consumer) => (
          <StyledExpander
            {...withSuomifiDefaultProps(this.props)}
            consumer={consumer}
          />
        )}
      </ExpanderGroupConsumer>
    );
  }
}

export { ExpanderConsumer };
