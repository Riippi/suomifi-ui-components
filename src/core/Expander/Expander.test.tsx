import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { axeTest } from '../../utils/test/axe';

import { Expander, ExpanderProps } from './Expander';
import { cssFromBaseStyles } from '../utils';
import { baseStyles } from './Expander.baseStyles';

const TestExpanderWithProps = (props: ExpanderProps, content: string) => {
  const { title, ...passProps } = props;
  return (
    <Expander title={title} {...passProps}>
      {content}
    </Expander>
  );
};

const TestExpander = TestExpanderWithProps(
  {
    title: 'Test expander',
    titleProps: { 'data-testid': 'expander-title' },
    className: 'expander-test',
  },
  'Test expander content',
);
describe('Basic expander', () => {
  it('render with the same component on the same container does not remount', () => {
    const expanderRenderer = render(TestExpander);
    const { getByTestId, rerender } = expanderRenderer;
    expect(getByTestId('expander-title').textContent).toBe('Test expander');

    // re-render the same component with different props
    rerender(
      TestExpanderWithProps(
        {
          title: 'Test expander two',
          titleProps: { 'data-testid': 'expander-title-2' },
          className: 'expander-test',
        },
        'Test expander content',
      ),
    );
    expect(getByTestId('expander-title-2').textContent).toBe(
      'Test expander two',
    );
  });

  it('shoud match snapshot', () => {
    const expanderRenderer = render(TestExpander);
    const { container } = expanderRenderer;
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('defaultOpen', () => {
  it('gives the classname to expander title and icon', () => {
    const { getByTestId } = render(
      TestExpanderWithProps(
        {
          title: 'Test expander open by default',
          titleProps: { 'data-testid': 'expander-open-by-default-title' },
          className: 'expander-open-by-default-test',
          defaultOpen: true,
        },
        'Test expander open by default content',
      ),
    );

    const button = getByTestId('expander-open-by-default-title');
    expect(button.classList.contains('fi-expander_title--open')).toBe(true);

    expect(
      button
        .querySelector('svg')
        ?.classList.contains('fi-expander_title-icon--open'),
    ).toBe(true);
  });

  it('classnames will be removed when clicked', () => {
    const mockClickHandler = jest.fn();
    const { getByTestId } = render(
      TestExpanderWithProps(
        {
          title: 'Test expander open by default',
          titleProps: { 'data-testid': 'expander-open-by-default-title' },
          className: 'expander-open-by-default-test',
          defaultOpen: true,
          onClick: mockClickHandler,
        },
        'Test expander open by default content',
      ),
    );

    const buttonToClick = getByTestId('expander-open-by-default-title');
    fireEvent.mouseDown(buttonToClick);

    expect(buttonToClick.classList.contains('fi-expander_title--open')).toBe(
      false,
    );

    expect(
      buttonToClick
        .querySelector('svg')
        ?.classList.contains('fi-expander_title-icon--open'),
    ).toBe(false);
  });
});

describe('onClick', () => {
  it('is called', async () => {
    const mockClickHandler = jest.fn();
    const { getByTestId } = render(
      TestExpanderWithProps(
        {
          title: 'Test expander onClick testing',
          titleProps: {
            'data-testid': 'expander-onclick-testing-title',
            onClick: mockClickHandler,
          },
          className: 'expander-onclick-test',
        },
        'Test expander click testing content',
      ),
    );
    const button = getByTestId('expander-onclick-testing-title');
    fireEvent.mouseDown(button);
    expect(mockClickHandler).toHaveBeenCalledTimes(1);
  });
});

describe('open', () => {
  it('open-classnames should be found ', async () => {
    const { getByTestId } = render(
      TestExpanderWithProps(
        {
          title: 'Test expander onClick testing',
          titleProps: {
            'data-testid': 'expander-onclick-testing-title',
          },
          className: 'expander-onclick-test',
          open: true,
        },
        'Test expander click testing content',
      ),
    );
    const button = getByTestId('expander-onclick-testing-title');
    expect(button.classList.contains('fi-expander_title--open')).toBe(true);

    expect(
      button
        .querySelector('svg')
        ?.classList.contains('fi-expander_title-icon--open'),
    ).toBe(true);
  });

  it('is clicked. Should not change as it is controlled outside', async () => {
    const mockClickHandler = jest.fn();
    const { getByTestId } = render(
      TestExpanderWithProps(
        {
          title: 'Test expander onClick testing',
          titleProps: {
            'data-testid': 'expander-onclick-testing-title',
            onClick: mockClickHandler,
          },
          className: 'expander-onclick-test',
          open: true,
        },
        'Test expander click testing content',
      ),
    );
    const button = getByTestId('expander-onclick-testing-title');
    fireEvent.mouseDown(button);
    expect(mockClickHandler).toHaveBeenCalledTimes(1);
    expect(button.classList.contains('fi-expander_title--open')).toBe(true);

    expect(
      button
        .querySelector('svg')
        ?.classList.contains('fi-expander_title-icon--open'),
    ).toBe(true);
  });
});

test('CSS export', () => {
  const css = cssFromBaseStyles(baseStyles);
  expect(css).toEqual(expect.stringContaining('background-color'));
});

test('should not have basic accessibility issues', axeTest(TestExpander));
