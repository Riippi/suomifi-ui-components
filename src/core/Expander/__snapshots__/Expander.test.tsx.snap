// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calling render with the same component on the same container does not remount 1`] = `
.c3 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  overflow: visible;
  text-transform: none;
  -webkit-appearance: button;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline-block;
  max-width: 100%;
  cursor: pointer;
}

.c3:-moz-focusring {
  outline: 1px dotted ButtonText;
}

.c3::-moz-focus-inner {
  border-style: none;
  padding: 0;
}

.c3::-webkit-file-upload-button {
  -webkit-appearance: button;
  font: inherit;
}

.c3::-webkit-inner-spin-button {
  height: auto;
}

.c3::-webkit-outer-spin-button {
  height: auto;
}

.c0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c4.fi-button--disabled {
  cursor: not-allowed;
}

.c1 {
  display: block;
  width: 100%;
  max-width: 100%;
}

.c7 {
  display: none;
}

.c5,
.c5 * {
  cursor: pointer;
}

.c5:hover,
.c5:active,
.c5:focus,
.c5:focus-within {
  cursor: pointer;
}

.c6 {
  display: inline-block;
  vertical-align: baseline;
}

.c2 {
  color: hsl(0,0%,16%);
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 18px;
  line-height: 1.5;
  font-weight: 400;
  background-color: hsl(0,0%,100%);
  position: relative;
  position: relative;
  padding: 0;
  border-radius: 2px;
  box-shadow: 0 1px 2px 0 rgba(41,41,41,0.14),0 1px 5px 0 rgba(41,41,41,0.12);
}

.c2:before {
  content: '';
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  width: 100%;
  height: 100%;
}

.c2:before {
  background-color: hsl(212,63%,98%);
  opacity: 0;
}

.c2 .fi-expander_title {
  color: hsl(0,0%,16%);
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 600;
  font-size: 14px;
  line-height: 20px;
  position: relative;
  display: block;
  width: 100%;
  font-size: font-family:'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 18px;
  line-height: 1.5;
  font-weight: 600;
  min-height: 60px;
}

.c2 .fi-expander_title:focus {
  outline: 0;
}

.c2 .fi-expander_title:focus:after {
  content: '';
  position: absolute;
  pointer-events: none;
  top: -2px;
  right: -2px;
  bottom: -2px;
  left: -2px;
  border-radius: 2px;
  background-color: transparent;
  border: 0px solid hsl(0,0%,100%);
  box-sizing: border-box;
  box-shadow: 0 0 0 2px hsl(196,77%,55%);
  z-index: 9999;
}

.c2 .fi-expander_title--no-tag {
  padding: 17px 60px 16px 20px;
  color: hsl(212,63%,45%);
}

.c2 .fi-expander_title-icon {
  position: absolute;
  height: 20px;
  width: 20px;
  top: 0;
  right: 0;
  margin: 20px;
}

.c2 .fi-expander_title--open .fi-expander_title-icon,
.c2 .fi-expander_title-icon--open {
  -webkit-transform: rotate(-180deg);
  -ms-transform: rotate(-180deg);
  transform: rotate(-180deg);
}

.c2 .fi-expander_title {
  color: hsl(0,0%,16%);
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 600;
  font-size: 14px;
  line-height: 20px;
  position: relative;
  display: block;
  width: 100%;
  font-size: font-family:'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 18px;
  line-height: 1.5;
  font-weight: 600;
  min-height: 60px;
}

.c2 .fi-expander_title--no-tag {
  padding: 17px 60px 16px 20px;
  color: hsl(212,63%,45%);
}

.c2 .fi-expander_title-icon {
  position: absolute;
  height: 20px;
  width: 20px;
  top: 0;
  right: 0;
  margin: 20px;
}

.c2 .fi-expander_title--open .fi-expander_title-icon,
.c2 .fi-expander_title-icon--open {
  -webkit-transform: rotate(-180deg);
  -ms-transform: rotate(-180deg);
  transform: rotate(-180deg);
}

.c2 > .fi-expander_content {
  position: relative;
  display: block;
  height: 0;
  overflow: hidden;
  word-break: break-word;
  -webkit-transform: scaleY(0);
  -ms-transform: scaleY(0);
  transform: scaleY(0);
  -webkit-transform-origin: top;
  -ms-transform-origin: top;
  transform-origin: top;
  -webkit-transition: all 50ms cubic-bezier(0.28,0.84,0.42,1);
  transition: all 50ms cubic-bezier(0.28,0.84,0.42,1);
  will-change: transition,height;
}

.c2 > .fi-expander_content:not(.fi-expander_content--no-padding) {
  padding: 0 16px;
}

.c2 > .fi-expander_content.fi-expander_content--open {
  height: 10%;
  overflow: visible;
  -webkit-animation: fi-expander_content-anim 50ms cubic-bezier(0.28,0.84,0.42,1) 1 forwards;
  animation: fi-expander_content-anim 50ms cubic-bezier(0.28,0.84,0.42,1) 1 forwards;
}

.c2 > .fi-expander_content.fi-expander_content--open:not(.fi-expander_content--no-padding) {
  padding-top: 0;
  padding-right: 20px;
  padding-bottom: 20px;
  padding-left: 20px;
}

<div
  class="c0 c1 c2 expander-test fi-expander"
>
  <button
    aria-disabled="false"
    aria-expanded="false"
    class="c3 c4 fi-button c5 fi-expander_title fi-expander_title--no-tag"
    data-testid="expander-title"
    tabindex="0"
    type="button"
  >
    Test expander
    <svg
      aria-hidden="true"
      class="fi-icon c6 fi-expander_title-icon"
      fill="hsl(212, 63%, 45%)"
      focusable="false"
      height="1em"
      viewBox="0 0 24 24"
      width="1em"
    >
      <path
        d="M13 15.6c-.3.3-.6.4-1 .4s-.7-.1-1-.4l-5.6-5.3c-.5-.5-.5-1.4 0-1.9s1.4-.5 2 0l4.6 4.4 4.6-4.4c.5-.5 1.4-.5 2 0 .5.5.5 1.4 0 1.9L13 15.6z"
      />
    </svg>
  </button>
  <div
    aria-hidden="true"
    class="c0 c1 c7 fi-expander_content"
    hidden=""
  >
    Test expander content
  </div>
</div>
`;
